### ISSUE TITLE
Cisco 7912 Connection Help

### ISSUE THREAD

**Issue by Ahh6483** [marcelloceschia/chan-sccp-b#568]
Hello there! I've got all my phones connection to the chan-sccp server working amazingly. I am having some trouble configuring the Cisco 7912, could someone help me out?
I'm struggling flashing the sccp firmware to the phone.

**Comment by dkgroot** [comment]
@Ahh6483 I have no direct experience with setting up a 7912 phone myself. Can you tell me how far you have gotten ?
Do you have the `CP7912080004SCCP080108A.sbin` firmware file ? Is the phone still booting ? Is it requesting ip-address from DHCP ? Is it trying to retrieve some firmware files or non at all via tftp ?

Do you have a XMLDefault.cnf.xml file containing a line for the 7912 to find it's firmware (amongst other lines):
```
<?xml version="1.0" encoding="UTF-8"?>
<Default>
...
    <loadInformation30007 model="Cisco IP Phone 7912">CP7912080004SCCP080108A</loadInformation30007>
...
</Default>
```
(Example: [XMLDefault.cnf.xml](https://github.com/dkgroot/provision_sccp/blob/master/tftpboot/settings/XMLDefault.cnf.xml) )

Also do re-ask the question on our [gitter chat](https://gitter.im/chan-sccp/chan-sccp) someone there might have one of these phone and gotten it running.

**Comment by Ahh6483** [comment]
Thank you for your reply!
Sure, I first had the 7912 running SIP, before I found the chan-sccp. I am unable to get it into the mode where it will retrieve the firmware files. I've tried to reset the phone many times, it gets assigned an IP address from the DHCP server, but it is not looking for the files. In the TFTP root is CP7912080004SCCP080108A.sbin, XMLDefault.cnf.xml, gk0014f28dd38a, and the gkdefault.cfg. 

I also have a 8961, I see its in the Advance Server Configuration however I cannot find much information on it. Is there a way to configure the 8961 with the chan-sccp?

**Comment by dkgroot** [comment]
@Ahh6483
Ok so your XMLDefault.cnf.xml should contain the firmware reference from above, so the phone knows that to load. 

## gkdefault
the gkdefault.txt file would have to contain something like:
```
#txt
# upgradecode:3,phonetype,0x0400,0x0100,tftp_server_ip,69,image_id,image_file_name
# phonetype: is 0x401 for the Cisco Unified IP Phone 7902G, 0x501 for the Cisco Unified IP Phone 7905G, or 0x601 for the Cisco Unified IP Phone 7912G.
# image_id is a unique 32-bit integer value. This value must be different for each upgrade. For this value, it is convenient to use 0x followed by the build date on the image file name. (For example, if the image file name is
# update 192.168.x.x to the tftp server ip-address you are using
upgradecode:3,0x601,0x0400,0x0100,192.168.x.x,69,0x170409a,CP7912080004SCCP080108A.sbin

# Set the pincode for admin access on the phone
UIPassword:1111

# You can start `prserv 9100` on your tftp server to monitor phones progress
NPrintf: 192.168.x.x.9100
```
And then you need to use cfgfmt.exe or ./cfgfmt.linux to convert this gkdefault.txt file to the actual gkdefault.cfg file.
```
./cfgfmt.linux -tsip_ptag.dat gkdefault.txt gkdefault.cfg
```

The 7912 phone series is a little different in this compared to other (newer) cisco models. Getting the gkdefault.txt file correct and converted to cfg is the crucial part. You can check in the tftp server log if the file is actually downloading these files.


See: [Cisco 7905g_7912g](https://www.cisco.com/en/US/docs/voice_ip_comm/cuipph/7905g_7912g/4_2/english/administration/guide/LowPalt.html#wpxref63598).

## Debug Log
To make it a little easier to see what is going on you can include:
```
NPrintf: 192.168.x.x.9100
```
In the gkdefault.txt file and run `prserv 9100` on you host server. This would force the phone to sent debug logging to the host server.

## Unlocking and preparing the Phone

The Cisco phone may have to be unlocked. This is accomplished by first navigating to Settings and from there to Network Configuration. In this menu, the unlock code must be entered. This is accomplished by typing **# followed by the password.

The password can be an alphanumeric string of up to nine characters. If it is set to 0 then the phone does not prompt for a password. However, the phone's configuration cannot be modified over the Web if the password is set to 0.

If the password is not known then the phone can be unlocked by restoring the factory settings. This is done as follows:

- Press *, 0 and softkey number 4 simultaneously. The manufacturing test-option screen should appear.
- Type 322873738# and press *.
- Press *, 0 and softkey number 4 simultaneously again, or power-cycle the phone. The phone should return to normal operation.

The phone can then be unlocked with the default password, which is 1234.

Once unlocked, press the Menu key, select Settings and Network Configuration. Choose the TFTP Enabled menu option and set the value to Yes. 

Firmware: [7912 firmware](https://github.com/dkgroot/provision_sccp/tree/master/tftpboot/firmware/7912)
Tools: [tools](https://github.com/dkgroot/provision_sccp/tree/master/tools/ATA-186_188)

**Comment by dkgroot** [comment]
Would you mind writing a little wiki page about the 7912 process, once you get it up and running ?

**Comment by evazzoler** [comment]
It seems that 7912 don't like the config ti be repeated, specially the one with the callmanager IP. It reboots in a loop... Just remove it from SEP<MAC>.XML...
And about gkdefault.cfg it must contain upgrade info and connect mode only. Search chan-sccp wiki for 7912 for the correct config.
You will reset the phone (search the **2 exact procedure) in order to let it read the gkdefault.cfg and it will read it if it won't find any other config file (please rename them or temporary move them). 
Good luck!

**Comment by dkgroot** [comment]
@Ahh6483 Did @evazzoler comments help ?

**Comment by Ahh6483** [comment]
Yes they did! After a little bit of messing around with TFTPd I was able to get it successfully connected. 

Thank you all!

**Comment by dkgroot** [comment]
@Ahh6483 Would it be worth a little hardware 7912 article on our wiki ? Or was it just a case of RTFM ?

**Comment by samam9911** [comment]
I am in same situation where the phone is going into loop mostly stuck at configuring IP. DHCP server shows and IP assigned and the tftp server shows: Jan 10 10:17:10 freepbx.sangoma.local in.tftpd[25712]: RRQ from 192.168.20.129 filename SEP001646CB4791.cnf.xml
but still unable to connect to the server. Would you please elaborate on what changes you made to tftpd to get it to work? Thanks!
